### ISSUE TITLE
Incorrect unused_variables diagnostic for or-ed patterns

### ISSUE THREAD

**Issue by sapir** [rust-lang/rust#67691]
With rustc 1.42.0-nightly (0de96d37f 2019-12-19):

For the following code:

```rust
pub enum MyEnum {
    A { i: i32, j: i32 },
    B { i: i32, j: i32 },
}

pub fn foo(x: MyEnum) {
    use MyEnum::*;

    match x {
        A { i, j } | B { i, j } => {
            println!("{}", i);
        }
    }
}
```

I get the following diagnostic:
```
) rustc +nightly test.rs -o test --crate-type lib
warning: unused variable: `j`
  --> test.rs:10:16
   |
10 |         A { i, j } | B { i, j } => {
   |                ^            ^
   |
   = note: `#[warn(unused_variables)]` on by default
help: try ignoring the field
   |
10 |         A { i, j: _j: _ } | B { i, j } => {
   |                ^^^^^^^
```

Instead of:
```
help: try ignoring the field
   |
10 |         A { i, j: _ } | B { i, j: _ } => {
   |                ^^^^            ^^^^
```

**Comment by Centril** [comment]
Regression introduced in 1.36.0.

**Comment by sapir** [comment]
I found the breaking commit with cargo-bisect-rustc and git bisect. https://github.com/rust-lang/rust/pull/59993

Based on this, I found that if the or pattern uses `ref` then the diagnostic becomes more confused.

I can have a try at fixing this later, if that's ok.

**Comment by Centril** [comment]
Was fixed in https://github.com/rust-lang/rust/pull/67766.
